lib/ns.c

```diff
@@ -185,6 +263,155 @@ static int ns_is_enabled(struct ns *ns)
 		return ns && ns->fd == -2 && ns->ns_id == NS_DEFAULT;
 }
 
+/* in case NEWNSID not available, the NSID will be locally obtained
+ */
+#define NS_BASE_NSID 200
+static ns_id_t ns_get_internal_nsid_fallback(int fd)
+{
+	return NS_BASE_NSID + fd;
+}
+
+#if defined(HAVE_NETLINK) && defined(VRF_BACKEND_NETNS_AVAILABLE)
+static int ns_temp_nsid;
+
+static int data_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+	/* skip unsupported attribute in user-space */
+	if (mnl_attr_type_valid(attr, IFA_MAX) < 0)
+		return MNL_CB_OK;
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+
+static int data_cb(const struct nlmsghdr *nlh, void *data)
+{
+	struct nlattr *tb[NETNSA_MAX+1] = {};
+
+	mnl_attr_parse(nlh, sizeof(struct rtgenmsg), data_attr_cb, tb);
+	if (tb[NETNSA_NSID])
+		ns_temp_nsid = mnl_attr_get_u32 (tb[NETNSA_NSID]);
+	return MNL_CB_OK;
+}
+
+/* values expected : not NETNSA_NSID_NOT_ASSIGNED */
+static ns_id_t ns_get_internal_nsid(int fd)
+{
+	struct mnl_socket *nl;
+	char buf[MNL_SOCKET_BUFFER_SIZE];
+	struct nlmsghdr *nlh;
+	struct rtgenmsg *rt;
+	int ret;
+	unsigned int seq, portid;
+	int ns_id = -1;
+
+	/* initialise nsid */
+	ns_temp_nsid = NS_UNKNOWN;
+	/* prepare netlink context */
+	nl = mnl_socket_open(NETLINK_ROUTE);
+	if (nl == NULL) {
+		zlog_err("netlink( %u) socket() error: %s",
+			 fd, safe_strerror(errno));
+		return ns_temp_nsid;
+	}
+
+	if (mnl_socket_bind(nl, RTNLGRP_NSID, MNL_SOCKET_AUTOPID) < 0) {
+		zlog_err("netlink( %u) socket() bind error: %s",
+			 fd, safe_strerror(errno));
+		mnl_socket_close(nl);
+		return ns_temp_nsid;
+	}
+
+	portid = mnl_socket_get_portid(nl);
+	/* emission and reception */
+
+	/* fill in context */
+	bzero(buf, sizeof(buf));
+	nlh = mnl_nlmsg_put_header(buf);
+	nlh->nlmsg_type = RTM_NEWNSID;
+	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK;
+	nlh->nlmsg_seq = seq = time(NULL);
+	rt = mnl_nlmsg_put_extra_header(nlh, sizeof(struct rtgenmsg));
+	rt->rtgen_family = AF_UNSPEC;
+
+	mnl_attr_put_u32(nlh, NETNSA_FD, fd);
+	mnl_attr_put_u32(nlh, NETNSA_NSID, ns_id);
```

Comment:
`NETNSA_NSID` was only introduced in the kernel by 2015. Will this VRF netns backend work on older kernel that don't have this capability?

Reply:
a fallback mecahnism is put in place, so that NSID will be locally generated, and so that the VRFID will be local to frr.